### DISCUSSION TITLE
it.only gives "Cannot read properties of undefined (reading 'parent')" with Stack Overflow solution for soft assertion

### DISCUSSION THREAD

**Discussion by MikeMcC399** [cypress-io/cypress#23280] [Questions and Help] [answered]
[Stack Overflow: "Does cypress support soft assertion?"](https://stackoverflow.com/questions/55868107/does-cypress-support-soft-assertion/55879165) which was published over 3 years ago suggests a solution for soft assertion.

At the time the contribution was published, `it.only` originally worked with Cypress 3.8.3 and excluded other tests from running. Since that time, attempting to use `it.only` causes an error (`Cannot read properties of undefined (reading 'parent')`) using the Stack Overflow code starting with Cypress 4.0.0 and including the latest Cypress 10.11.0.

I asked in July 2022 in the above [Stack Overflow article](https://stackoverflow.com/questions/55868107/does-cypress-support-soft-assertion/55879165) if there is a solution to allow `it.only` to work with this suggested code. There hasn't been any solution there yet.

Can anybody advise what needs to be changed to get `it.only` to work in this environment?

The [Stack Overflow post: How can I make `.only` and `this` to work after hacking the Cypress hook?](https://stackoverflow.com/questions/72111937/how-can-i-make-only-and-this-to-work-after-hacking-the-cypress-hook) also didn't receive any solution for `it.only` so far.

## Steps to reproduce

To reproduce, take the code from [Stack Overflow: "Does cypress support soft assertion?"](https://stackoverflow.com/questions/55868107/does-cypress-support-soft-assertion/55879165) and add it into `support/index.js` (Cypress 9 and earlier) or into `support/e2e.js` (Cypress 10), then execute the simple Cypress test spec:

```js
describe('it.only test', () => {
    it.only('test', () => {
        cy.log("Printing a message to the Cypress Command Log.")
    });
});
```

The console logs shows the following on Cypress 9.7.0:

```
TypeError: The following error originated from your test code, not from Cypress.

  > Cannot read properties of undefined (reading 'parent')

When Cypress detects uncaught errors originating from your test code it will automatically fail the current test.

Cypress could not associate this error to any specific test.

We dynamically generated a new test to display this failure.
    at Object.only (http://localhost:63799/__cypress/runner/cypress_runner.js:125030:14)
    at Function.context.it.only (http://localhost:63799/__cypress/runner/cypress_runner.js:124834:26)
    at Function.only (http://localhost:63799/__cypress/runner/cypress_runner.js:183242:21)
    at Object.create (http://localhost:63799/__cypress/runner/cypress_runner.js:125002:19)
    at context.describe.context.context (http://localhost:63799/__cypress/runner/cypress_runner.js:124778:27)
    at http://localhost:63799/__cypress/runner/cypress_runner.js:183242:21
    at http://localhost:63799/__cypress/runner/cypress_runner.js:186274:16
    at arrayEach (http://localhost:63799/__cypress/runner/cypress_runner.js:41318:11)
    at Function.forEach (http://localhost:63799/__cypress/runner/cypress_runner.js:50198:14)
    at evalScripts (http://localhost:63799/__cypress/runner/cypress_runner.js:186273:49)
    at http://localhost:63799/__cypress/runner/cypress_runner.js:186281:152
From previous event:
    at runScriptsFromUrls (http://localhost:63799/__cypress/runner/cypress_runner.js:186281:136)
    at Object.runScripts (http://localhost:63799/__cypress/runner/cypress_runner.js:186296:12)
    at $Cypress.onSpecWindow (http://localhost:63799/__cypress/runner/cypress_runner.js:174692:76)
```

or after migrating to Cypress 10.4.0, where the error message is identical to 9.7.0 and the stack trace is also the same, apart from line numbers:

```
TypeError: The following error originated from your test code, not from Cypress.

  > Cannot read properties of undefined (reading 'parent')

When Cypress detects uncaught errors originating from your test code it will automatically fail the current test.

Cypress could not associate this error to any specific test.

We dynamically generated a new test to display this failure.
    at Object.only (http://localhost:62671/__cypress/runner/cypress_runner.js:118180:14)
    at context.it.only (http://localhost:62671/__cypress/runner/cypress_runner.js:117984:26)
    at Function.only (http://localhost:62671/__cypress/runner/cypress_runner.js:176870:21)
    at Object.create (http://localhost:62671/__cypress/runner/cypress_runner.js:118152:19)
    at context.describe.context.context (http://localhost:62671/__cypress/runner/cypress_runner.js:117928:27)
    at http://localhost:62671/__cypress/runner/cypress_runner.js:176870:21
    at http://localhost:62671/__cypress/runner/cypress_runner.js:179915:16
    at arrayEach (http://localhost:62671/__cypress/runner/cypress_runner.js:40671:11)
    at Function.forEach (http://localhost:62671/__cypress/runner/cypress_runner.js:49551:14)
    at evalScripts (http://localhost:62671/__cypress/runner/cypress_runner.js:179914:49)
    at http://localhost:62671/__cypress/runner/cypress_runner.js:179922:152
From previous event:
    at runScriptsFromUrls (http://localhost:62671/__cypress/runner/cypress_runner.js:179922:136)
    at Object.runScripts (http://localhost:62671/__cypress/runner/cypress_runner.js:179937:12)
    at $Cypress.onSpecWindow (http://localhost:62671/__cypress/runner/cypress_runner.js:168145:76)
```

## Expectation

Attempting to run the test spec above should run the test without any error message when using the code from [Stack Overflow: "Does cypress support soft assertion?"](https://stackoverflow.com/questions/55868107/does-cypress-support-soft-assertion/55879165).

Under the old Cypress 3.8.3 version this is no problem:

![only-3 8 3](https://user-images.githubusercontent.com/66998419/184165496-b7f22e36-da9a-490a-b166-5f42f88f6f7f.jpg)

**Comment by chip-ext-star** [comment]
hi @MikeMcC399.

I found a workaround:

```
window.it.only = (title, func) => {
  let cases = Cypress.env('CypressOnly');
  cases += ';';
  cases += title;
  Cypress.env('CypressOnly', cases);
  origIt(title, func && (() => itCallback(title, func)));
};

const filterTestCases = () => {
  const testFilter = Cypress.env('CypressOnly')?.split(';');
  let tests = Cypress.mocha.getRunner().suite.suites?.[0].tests;
  tests.forEach((test, index) => {
    if (testFilter?.length > 1 && !testFilter?.some(t => t.includes(test.title)))
      Cypress.mocha.getRunner().suite.suites[0].tests[index].pending = true;
  });
};

before(() => {
    filterTestCases();
})
```

Hope this helps you.

**Reply by MikeMcC399** [reply]
@chip-ext-star 

Thank you very much for your response and solution suggestion. I replaced

```
window.it.only = (title, func) => {
    origIt.only(title, func && (() => itCallback(func)));
};
```
with your code and ran the test
```
describe('it.only test', () => {
    it.only('test', () => {
        cy.log("Printing a message to the Cypress Command Log.")
    });
});
```
using Cypress 9.7.0 and I got a different error message:

![image](https://user-images.githubusercontent.com/66998419/192129760-1993b31b-8846-496b-9614-1421dfd95b3a.png)

Unfortunately I don't have the skills to debug this further. Perhaps you have an idea?

**Reply by thiagocosenza** [reply]
Hi @MikeMcC399 , how did you implement the solution of [(https://stackoverflow.com/questions/55868107/does-cypress-support-soft-assertion)] using Cypress >10.0? I didnt manage to do it, because index.js has gone.

**Reply by MikeMcC399** [reply]
@thiagocosenza 
> how did you implement the solution of [(https://stackoverflow.com/questions/55868107/does-cypress-support-soft-assertion)] using Cypress >10.0? I didnt manage to do it, because index.js has gone.

I installed on Cypress 9 and then used the migration tool for Cypress 10. This renames `support/index.js` to `support/e2e.js`. See https://docs.cypress.io/guides/references/migration-guide#supportFile.

**Reply by thiagocosenza** [reply]
Good! Thank you.  And did you manage to fix the it.only afterwards?

**Reply by MikeMcC399** [reply]
> And did you manage to fix the it.only afterwards?

No. This issue is still unresolved.

**Reply by DPelyp** [reply]
> hi @chip-ext-star
> 
> I found a workaround:
> 
> ```
> window.it.only = (title, func) => {
>   let cases = Cypress.env('CypressOnly');
>   cases += ';';
>   cases += title;
>   Cypress.env('CypressOnly', cases);
>   filterTestCases();
>   origIt(title, func && (() => itCallback(title, func)));
> };
> 
> const filterTestCases = () => {
>   const cypressOnly = Cypress.env('CypressOnly')?.split(';');
>   if (!cypressOnly) return;
>   let tests = Cypress.mocha.getRunner().suite.suites?.[0].tests;
>   tests.forEach((test, index) => {
>     if (cypressOnly.indexOf(test.title) < 0)
>       Cypress.mocha.getRunner().suite.suites[0].tests[index].pending = true;
>   });
> };
> ```
> 
> Hope this helps you.

Didn't get how to fix the cy.only issue, test title should be added into env variable? Maybe you have example how to work with that? I will be very grateful

**Reply by MikeMcC399** [reply]
Hi again @chip-ext-star !

It looks like
`origIt(title, func && (() => itCallback(title, func)));` should be
`origIt(title, func && (() => itCallback(func)));` which allows `it.only` to work.

If you would like to check the above and modify your suggested code then I would set the discussion to "answered".

**Comment by MikeMcC399** [comment]
- I wonder if Gleb @bahmutov might weigh in on this issue? He commented on the same code in https://github.com/cypress-io/cypress/issues/17300#issuecomment-879127820

- The original problematic code for `support/e2e.js` is on [Stack Overflow: "Does cypress support soft assertion?"](https://stackoverflow.com/questions/55868107/does-cypress-support-soft-assertion/55879165) which causes a problem with `it.only`.

- `it.only` with this monkey patch stopped working with the Cypress 4.0.0 release which was the first release to allow multiple tests using `it.only`. In Cypress v3 it still worked.

In addition there is a new issue:
- `it.skip` stopped working in conjunction with the monkey-patch after updating Cypress 10.10.0 to Cypress 10.11.0
it now produces the error message: "Cannot set properties of undefined (setting 'body')" logged under https://github.com/cypress-io/cypress/issues/24525 (support was declined for this issue and the issue is closed unresolved).

**Comment by chip-ext-star** [comment, marked as answer]
hi,

I make an example here: https://github.com/lehoanglu0ng/cypress-only

**Reply by MikeMcC399** [reply]
Hi @chip-ext-star 

Thanks for your repo with the fix for `it.only`. I confirm that your test file works.

It would be great if it also worked with a test containing `it.skip`, although that wasn't part of the original discussion.
If I run a test with `it.skip` and with your modified monkey-patch I get the error:
```
TypeError: Cannot read properties of undefined (reading 'tests')

Because this error occurred during a `before all` hook we are skipping all of the remaining tests.
    at filterTestCases (webpack:///./cypress/support/e2e.js:105:58)
    at Context.eval (webpack:///./cypress/support/e2e.js:100:2)

```

**Reply by chip-ext-star** [reply]
Happy to hear it helps you. I will find the way to make `it.skip` works.
